File: asherah-dotnet-shared/src/Crypto/BasicExpiringCryptoPolicy.cs
using System;

namespace GoDaddy.Asherah.Crypto;

public class BasicExpiringCryptoPolicy : CryptoPolicy
{
    private readonly long keyExpirationMillis;
    private readonly long revokeCheckMillis;

    private readonly KeyRotationStrategy keyRotationStrategy;
    private readonly bool canCacheSystemKeys;
    private readonly bool canCacheIntermediateKeys;
    private readonly bool canCacheSessions;
    private readonly long sessionCacheMaxSize;
    private readonly long sessionCacheExpireMillis;
    private readonly bool notifyExpiredSystemKeyOnRead;
    private readonly bool notifyExpiredIntermediateKeyOnRead;

    private BasicExpiringCryptoPolicy(Builder builder)
    {
        keyExpirationMillis = (long)TimeSpan.FromDays(builder.KeyExpirationDays).TotalMilliseconds;
        revokeCheckMillis = (long)TimeSpan.FromMinutes(builder.RevokeCheckMinutes).TotalMilliseconds;
        keyRotationStrategy = builder.KeyRotationStrategy;
        canCacheSystemKeys = builder.CanCacheSystemKeys;
        canCacheIntermediateKeys = builder.CanCacheIntermediateKeys;
        canCacheSessions = builder.CanCacheSessions;
        sessionCacheMaxSize = builder.SessionCacheMaxSize;
        sessionCacheExpireMillis = builder.SessionCacheExpireMillis;
        notifyExpiredSystemKeyOnRead = builder.NotifyExpiredSystemKeyOnRead;
        notifyExpiredIntermediateKeyOnRead = builder.NotifyExpiredIntermediateKeyOnRead;
    }

    public interface IKeyExpirationDaysStep
    {
        IRevokeCheckMinutesStep WithKeyExpirationDays(int days);
    }

    public interface IRevokeCheckMinutesStep
    {
        IBuildStep WithRevokeCheckMinutes(int minutes);
    }

    public interface IBuildStep
    {
        IBuildStep WithRotationStrategy(KeyRotationStrategy rotationStrategy);
        IBuildStep WithCanCacheSystemKeys(bool cacheSystemKeys);
        IBuildStep WithCanCacheIntermediateKeys(bool cacheIntermediateKeys);
        IBuildStep WithCanCacheSessions(bool cacheSessions);
        IBuildStep WithSessionCacheMaxSize(long sessionCacheMaxSize);
        IBuildStep WithSessionCacheExpireMillis(long sessionCacheExpireMillis);
        IBuildStep WithNotifyExpiredSystemKeyOnRead(bool notify);
        IBuildStep WithNotifyExpiredIntermediateKeyOnRead(bool notify);
        BasicExpiringCryptoPolicy Build();
    }

    public static IKeyExpirationDaysStep NewBuilder() => new Builder();

    public override bool IsKeyExpired(DateTimeOffset keyCreationDate)
    {
        long currentUnixTimeMillis = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        long expiredMillis = keyCreationDate.ToUnixTimeMilliseconds() + keyExpirationMillis;
        return currentUnixTimeMillis > expiredMillis;
    }

    public override long GetRevokeCheckPeriodMillis() => revokeCheckMillis;
    public override bool CanCacheSystemKeys() => canCacheSystemKeys;
    public override bool CanCacheIntermediateKeys() => canCacheIntermediateKeys;
    public override bool CanCacheSessions() => canCacheSessions;
    public override long GetSessionCacheMaxSize() => sessionCacheMaxSize;
    public override long GetSessionCacheExpireMillis() => sessionCacheExpireMillis;
    public override bool NotifyExpiredIntermediateKeyOnRead() => notifyExpiredIntermediateKeyOnRead;
    public override bool NotifyExpiredSystemKeyOnRead() => notifyExpiredSystemKeyOnRead;
    public override KeyRotationStrategy GetKeyRotationStrategy() => keyRotationStrategy;

    internal long KeyExpirationMillis => keyExpirationMillis;
    internal long RevokeCheckMillis => revokeCheckMillis;

    private class Builder : IKeyExpirationDaysStep, IRevokeCheckMinutesStep, IBuildStep
    {
        private int keyExpirationDays;
        private int revokeCheckMinutes;

        private KeyRotationStrategy keyRotationStrategy = DefaultKeyRotationStrategy;
        private bool canCacheSystemKeys = DefaultCanCacheSystemKeys;
        private bool canCacheIntermediateKeys = DefaultCanCacheIntermediateKeys;
        private bool canCacheSessions = DefaultCanCacheSessions;
        private long sessionCacheMaxSize = DefaultSessionCacheSize;
        private long sessionCacheExpireMillis = DefaultSessionCacheExpiryMillis;
        private bool notifyExpiredSystemKeyOnRead = DefaultNotifyExpiredSystemKeyOnRead;
        private bool notifyExpiredIntermediateKeyOnRead = DefaultNotifyExpiredIntermediateKeyOnRead;

        internal int KeyExpirationDays => keyExpirationDays;
        internal int RevokeCheckMinutes => revokeCheckMinutes;
        internal KeyRotationStrategy KeyRotationStrategy => keyRotationStrategy;
        internal bool CanCacheSystemKeys => canCacheSystemKeys;
        internal bool CanCacheIntermediateKeys => canCacheIntermediateKeys;
        internal bool CanCacheSessions => canCacheSessions;
        internal long SessionCacheMaxSize => sessionCacheMaxSize;
        internal long SessionCacheExpireMillis => sessionCacheExpireMillis;
        internal bool NotifyExpiredSystemKeyOnRead => notifyExpiredSystemKeyOnRead;
        internal bool NotifyExpiredIntermediateKeyOnRead => notifyExpiredIntermediateKeyOnRead;

        private const KeyRotationStrategy DefaultKeyRotationStrategy = KeyRotationStrategy.Inline;
        private const bool DefaultCanCacheSystemKeys = true;
        private const bool DefaultCanCacheIntermediateKeys = true;
        private const bool DefaultCanCacheSessions = false;
        private const long DefaultSessionCacheSize = 1000;
        private const long DefaultSessionCacheExpiryMillis = 120000;
        private const bool DefaultNotifyExpiredSystemKeyOnRead = false;
        private const bool DefaultNotifyExpiredIntermediateKeyOnRead = false;

        public IRevokeCheckMinutesStep WithKeyExpirationDays(int days)
        {
            keyExpirationDays = days;
            return this;
        }

        public IBuildStep WithRevokeCheckMinutes(int minutes)
        {
            revokeCheckMinutes = minutes;
            return this;
        }

        public IBuildStep WithRotationStrategy(KeyRotationStrategy rotationStrategy)
        {
            keyRotationStrategy = rotationStrategy;
            return this;
        }

        public IBuildStep WithCanCacheSystemKeys(bool cacheSystemKeys)
        {
            canCacheSystemKeys = cacheSystemKeys;
            return this;
        }

        public IBuildStep WithCanCacheIntermediateKeys(bool cacheIntermediateKeys)
        {
            canCacheIntermediateKeys = cacheIntermediateKeys;
            return this;
        }

        public IBuildStep WithCanCacheSessions(bool cacheSessions)
        {
            canCacheSessions = cacheSessions;
            return this;
        }

        public IBuildStep WithSessionCacheMaxSize(long sessionCacheMaxSize)
        {
            this.sessionCacheMaxSize = sessionCacheMaxSize;
            return this;
        }

        public IBuildStep WithSessionCacheExpireMillis(long sessionCacheExpireMillis)
        {
            this.sessionCacheExpireMillis = sessionCacheExpireMillis;
            return this;
        }

        public IBuildStep WithNotifyExpiredSystemKeyOnRead(bool notify)
        {
            notifyExpiredSystemKeyOnRead = notify;
            return this;
        }

        public IBuildStep WithNotifyExpiredIntermediateKeyOnRead(bool notify)
        {
            notifyExpiredIntermediateKeyOnRead = notify;
            return this;
        }

        public BasicExpiringCryptoPolicy Build() => new(this);
    }
}

File: asherah-java/java/src/main/java/com/godaddy/asherah/crypto/BasicExpiringCryptoPolicy.java
package com.godaddy.asherah.crypto;

import com.godaddy.asherah.jni.AsherahConfig;

import java.time.Instant;
import java.time.temporal.ChronoUnit;

/**
 * Configurable expiring crypto policy. Maps expiry/cache settings to native config.
 * Compatible with the canonical godaddy/asherah BasicExpiringCryptoPolicy.
 */
public class BasicExpiringCryptoPolicy extends CryptoPolicy {

    private final long keyExpirationMillis;
    private final long revokeCheckPeriodMillis;
    private final KeyRotationStrategy rotationStrategy;
    private final boolean cacheSystemKeys;
    private final boolean cacheIntermediateKeys;
    private final boolean cacheSessions;
    private final long sessionCacheMaxSize;
    private final long sessionCacheExpireMillis;
    private final boolean notifyExpiredSystemKeyOnRead;
    private final boolean notifyExpiredIntermediateKeyOnRead;

    private BasicExpiringCryptoPolicy(final Builder builder) {
        this.keyExpirationMillis = builder.keyExpirationDays * 24L * 60 * 60 * 1000;
        this.revokeCheckPeriodMillis = builder.revokeCheckMinutes * 60L * 1000;
        this.rotationStrategy = builder.rotationStrategy;
        this.cacheSystemKeys = builder.cacheSystemKeys;
        this.cacheIntermediateKeys = builder.cacheIntermediateKeys;
        this.cacheSessions = builder.cacheSessions;
        this.sessionCacheMaxSize = builder.sessionCacheMaxSize;
        this.sessionCacheExpireMillis = builder.sessionCacheExpireMinutes * 60L * 1000;
        this.notifyExpiredSystemKeyOnRead = builder.notifyExpiredSystemKeyOnRead;
        this.notifyExpiredIntermediateKeyOnRead = builder.notifyExpiredIntermediateKeyOnRead;
    }

    @Override
    public void applyConfig(final AsherahConfig.Builder builder) {
        super.applyConfig(builder);
        builder.expireAfter(keyExpirationMillis / 1000);
        builder.checkInterval(revokeCheckPeriodMillis / 1000);
    }

    @Override
    public boolean isKeyExpired(final Instant keyCreationDate) {
        return keyCreationDate.plus(keyExpirationMillis, ChronoUnit.MILLIS).isBefore(Instant.now());
    }

    @Override
    public long getRevokeCheckPeriodMillis() {
        return revokeCheckPeriodMillis;
    }

    @Override
    public boolean canCacheSystemKeys() {
        return cacheSystemKeys;
    }

    @Override
    public boolean canCacheIntermediateKeys() {
        return cacheIntermediateKeys;
    }

    @Override
    public boolean canCacheSessions() {
        return cacheSessions;
    }

    @Override
    public long getSessionCacheMaxSize() {
        return sessionCacheMaxSize;
    }

    @Override
    public long getSessionCacheExpireMillis() {
        return sessionCacheExpireMillis;
    }

    @Override
    public boolean notifyExpiredIntermediateKeyOnRead() {
        return notifyExpiredIntermediateKeyOnRead;
    }

    @Override
    public boolean notifyExpiredSystemKeyOnRead() {
        return notifyExpiredSystemKeyOnRead;
    }

    @Override
    public KeyRotationStrategy keyRotationStrategy() {
        return rotationStrategy;
    }

    public static KeyExpirationDaysStep newBuilder() {
        return new Builder();
    }

    public interface KeyExpirationDaysStep {
        RevokeCheckMinutesStep withKeyExpirationDays(int days);
    }

    public interface RevokeCheckMinutesStep {
        BuildStep withRevokeCheckMinutes(int minutes);
    }

    public interface BuildStep {
        BuildStep withRotationStrategy(KeyRotationStrategy rotationStrategy);
        BuildStep withCanCacheSystemKeys(boolean cacheSystemKeys);
        BuildStep withCanCacheIntermediateKeys(boolean cacheIntermediateKeys);
        BuildStep withCanCacheSessions(boolean cacheSessions);
        BuildStep withSessionCacheMaxSize(long sessionCacheMaxSize);
        BuildStep withSessionCacheExpireMinutes(int sessionCacheExpireMinutes);
        BuildStep withNotifyExpiredSystemKeyOnRead(boolean notify);
        BuildStep withNotifyExpiredIntermediateKeyOnRead(boolean notify);
        BasicExpiringCryptoPolicy build();
    }

    public static final class Builder implements KeyExpirationDaysStep, RevokeCheckMinutesStep, BuildStep {
        private int keyExpirationDays;
        private int revokeCheckMinutes;
        private KeyRotationStrategy rotationStrategy = KeyRotationStrategy.INLINE;
        private boolean cacheSystemKeys = true;
        private boolean cacheIntermediateKeys = true;
        private boolean cacheSessions = false;
        private long sessionCacheMaxSize = 1000;
        private int sessionCacheExpireMinutes = 120;
        private boolean notifyExpiredSystemKeyOnRead = false;
        private boolean notifyExpiredIntermediateKeyOnRead = false;

        private Builder() {}

        @Override
        public RevokeCheckMinutesStep withKeyExpirationDays(final int days) {
            this.keyExpirationDays = days;
            return this;
        }

        @Override
        public BuildStep withRevokeCheckMinutes(final int minutes) {
            this.revokeCheckMinutes = minutes;
            return this;
        }

        @Override
        public BuildStep withRotationStrategy(final KeyRotationStrategy rotationStrategy) {
            this.rotationStrategy = rotationStrategy;
            return this;
        }

        @Override
        public BuildStep withCanCacheSystemKeys(final boolean cacheSystemKeys) {
            this.cacheSystemKeys = cacheSystemKeys;
            return this;
        }

        @Override
        public BuildStep withCanCacheIntermediateKeys(final boolean cacheIntermediateKeys) {
            this.cacheIntermediateKeys = cacheIntermediateKeys;
            return this;
        }

        @Override
        public BuildStep withCanCacheSessions(final boolean cacheSessions) {
            this.cacheSessions = cacheSessions;
            return this;
        }

        @Override
        public BuildStep withSessionCacheMaxSize(final long sessionCacheMaxSize) {
            this.sessionCacheMaxSize = sessionCacheMaxSize;
            return this;
        }

        @Override
        public BuildStep withSessionCacheExpireMinutes(final int sessionCacheExpireMinutes) {
            this.sessionCacheExpireMinutes = sessionCacheExpireMinutes;
            return this;
        }

        @Override
        public BuildStep withNotifyExpiredSystemKeyOnRead(final boolean notify) {
            this.notifyExpiredSystemKeyOnRead = notify;
            return this;
        }

        @Override
        public BuildStep withNotifyExpiredIntermediateKeyOnRead(final boolean notify) {
            this.notifyExpiredIntermediateKeyOnRead = notify;
            return this;
        }

        @Override
        public BasicExpiringCryptoPolicy build() {
            return new BasicExpiringCryptoPolicy(this);
        }
    }
}
